README.md
StackMob Sencha Touch SDK
=========================

This project is a (work in progress) SDK for connecting a Sencha Touch application with StackMob via their
OAuth 2.0 REST API.  The SDK is very much in its infancy, and is in a state of flux.  Consider this very
much a beta.

Installation
-----

1. Place the extension in your codebase (might we suggest `ux/stackmob-senchatouch`). Note that you will need to
modify your classpath in `sencha.cfg` if you locate the `ux` folder outside of your `app` directory.

1. Tell `Ext.Loader` (in `app.js`) where to find the extension's namespace:

    ```javascript
    Ext.Loader.setConfig({
        enabled: true,
        paths: {
            'Ext': '/touch/src',
            'Ux.palominolabs.stackmob': 'ux/stackmob-senchatouch'
        }
    });
    ```

1. Require the extension loader in your `Ext.Application`:

    ```javascript
    requires: ['Ux.palominolabs.stackmob.Loader'];
    ```

1. Supply your StackMob connection information in your `Ext.Application`'s configuration:

    ```javascript
    Ext.application({
        // ... snip ...
        stackMob: {
            appName: '<YOUR APP NAME>',
            publicKey: '<YOUR PUBLIC KEY>'
        },
        // ... snip ...
    });
    ```

    If you are accessing StackMob's API through a cross-origin domain, you will also need to add `enableCors: true` to the `init` config.

Usage
-----

### Stores and Models

To back an `Ext.Store` or `Ext.data.Model` with StackMob, simply use the `stackmob` proxy and supply the URL for the StackMob schema:
```javascript
Ext.define('YourAwesomeApp.store.NiftyStore', { // (or model)
    extend: 'Ext.data.Store', // (or Ext.data.Model)

    config: {
        model: 'YourAwesomeApp.model.NiftyThing',
        proxy: {
            type: 'stackmob',
            url: 'yourstackmobschemaname'
        }
    }
});
```

### AJAX

To make an AJAX call to StackMob without using an `Ext.Store`, simply use the `Ux.palominolabs.stackmob.StackMobAjax`
object wherever you would normally use `Ext.Ajax`:
```javascript
Ux.palominolabs.stackmob.StackMobAjax.request({
    url: 'yourstackmobobjectname',
    success: function(respose) {
        // ...
    },
    // ...
});
```

### Authentication

#### Logging In
To provide a login form for users to authenticate with StackMob, you can subclass
`Ux.palominolabs.stackmob.form.StackMobForm`.  This form will add the appropriate headers
and hidden fields for you, and handle the authentication logic on successful submission.  All you have
to do is add the required fields, and then calling `submit` on the form will Just Work:
```javascript
Ext.define("StackMobSenchaTouchDemo.view.account.Login", {
    extend: 'Ux.palominolabs.stackmob.form.StackMobForm',

    config: {
        items: [{
            xtype: 'fieldset',
            items: [{
                xtype: 'textfield',
                name: 'username',
                label: 'Username'
            },{
                xtype: 'passwordfield',
                name: 'password',
                label: 'Password'
            }]
        },{
            xtype: 'button',
            name: 'submit',
            action: 'login',
            text: 'Login'
        }]
    }
});
```

Note that the names of these fields should match the login field and password field defined in your
StackMob login schema.  The login schema itself is assumed to be `user`, but if you are using a
different schema name, you can customize it by specifying `loginSchema` in your call to `StackMobConnector.init`.

Once your user has logged in, you do not need to do anything special to support API calls that require
authentication.

#### Checking If a User Is Logged In

To check if a user is authenticated, call `Ux.palominolabs.stackmob.data.StackMobConnector.isAuthenticated()`.

#### Checking Which User Is Logged In

You can get the authenticated user's data by calling
`Ux.palominolabs.stackmob.data.StackMobConnector.getAuthenticatedUserData()`.  This will return an JavaScript
Object containing the fields for the StackMob object.

#### Logging Out

To logout, call `Ux.palominolabs.stackmob.data.StackMobConnector.logout()`.

#### Creating a New User

To create a new user, call `Ux.palominolabs.stackmob.data.StackMobConnector.createUser(options)`.
The options should include `userData` to populate the new user entry with. All other options available for
`Ext.data.Connection.request` are also supported.

#### Forgot Password

To send the user a temporary password, call `Ux.palominolabs.stackmob.data.StackMobConnector.forgotPassword(options)`.
The options should include a `username` and your `user` schema needs to have an `email address` field (this can be the
`username` field). StackMob will automatically send an email to the user with a temporary password. If they try to login
with this temporary password, it will fail with an `error_description` of `Temporary password reset required.`

To allow the user to create a new password using the temporary password, you can extend the same `StackMobForm`to
create a new login that takes a username, password, and new_password. See [StackMob's Forgot Password docs](https://developer.stackmob.com/rest-api/api-docs#a-forgot_password)
for more details.

#### Reset Password

To reset the logged-in user's password, call `Ux.palominolabs.stackmob.data.StackMobConnector.resetPassword(options)`.
The options should include both an `oldPassword` and a `newPassword`. All other options available for
`Ext.data.Connection.request` are also supported.

### Storing Files in S3

StackMob supports file uploads in the form of [binary fields](https://developer.stackmob.com/module/s3).  This SDK
provides a few tools for interacting with these fields, either as a model field, or directly in an AJAX request.  The
core of this functionality is provided by the `Ux.palominolabs.stackmob.util.File` class.

Given a native `File` object, you can read the file into an encoded string representation which can be used
in the DOM, as well as converted to the transport format required by StackMob.  To create the data URL, you can do
something like this:
```javascript
// Given a File object called myFile with filename myFileFilename...
var encodedDataUrl;
Ux.palominolabs.stackmob.util.File.readFile(myFile, function (intermediateDataUrl) {
    encodedDataUrl = Ux.palominolabs.stackmob.util.File.embedFilenameInDataUrl(myFileFilename, intermediateDataUrl);
});
```

...which results in a data URL which can be, for instance, used as the `src` attribute of an `img` tag, and can further
be converted into StackMob's required format for upload:
```javascript
var stackMobFormattedFile = Ux.palominolabs.stackmob.util.File.generateStackMobBinaryDataForDataUrlWithFilename(encodedDataUrl);
```

If you are executing a manual `StackMobAjax.request`, you can use `stackMobFormattedFile` to upload the file.  If,
however, the file is the value of a model field, the process is slightly simpler.  You can use the `stackmobbinary`
field data type to handle the final step (`generateStackMobBinaryDataForDataUrlWithFilename`) when necessary:
```javascript
Ext.define('YourAwesomeApp.model.NiftyModel', {
    extend: 'Ext.data.Model',

    config: {
        field: [
            {name: 'picture', type: 'stackmobbinary'}
        ]
        proxy: {
            type: 'stackmob',
            url: 'yourstackmobschemaname'
        }
    }
});
```

In this way, if you `set` the `picture` field to values which are generate by `Ux.palominolabs.stackmob.util.File.embedFilenameInDataUrl`
as demonstrated above, the `stackmob` proxy will automatically handle converting the data URL into the format expected
by StackMob.  This field type is polymorphic, in that "dirty" values that you set will be data URLs describing a file,
whereas "clean" values that come from StackMob will be S3 URLs.  In many cases, this will Just Work as expected,
but this is an important caveat to be aware of.  A working example of using a model with a `stackmobbinary` field
can be found in the example application.

Remember, to use binary fields, you must [configure your schema](https://developer.stackmob.com/module/s3) appropriately
in StackMob.

Example
-------

See the README in the `example` directory for instructions on running the included example application.

Quick Note: Connecting to StackMob
----------------------

Keep in mind that, due to cross-domain restrictions, in order to communicate with StackMob from your
Sencha Touch app in development, you may want to [enable CORS for your application](https://dashboard.stackmob.com/module/cors).

Also keep in mind that you must enable OAuth 2.0 for your StackMob application to use this extension.

Future Enhancements & Known Issues
----------------------------------

For the complete list (or to report a bug or request a feature), check out the
[issue tracker](https://github.com/tylerwolf/stackmob-senchatouch-sdk/issues), but the broad strokes are as follows:

- Filtering is coming soon.
- Support for associations is coming coon.

Contributing
------------

This SDK is an open source project, so bug reports are welcome and pull requests are encouraged.  We'd also love
to hear your feedback - be it questions, comments, concerns, worries, whimpers, or whines.  Or feature requests.

Copyright
---------

Copyright 2012-2013 Palomino Labs, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

This product also includes software developed by:

 - Sencha Inc. (http://www.sencha.com)
 - StackMob, Inc. (http://www.stackmob.com)
 - Jeff Mott (http://code.google.com/p/crypto-js/)

Changelog.md
HEAD
====

v0.4.1 - 2013/10/30
===================

- Added support for uploading binary files to S3.
- [#16] Added in forgot password functionality
- [#15] Updated example to demonstrate logging in with temporary password

v0.4.0 - 2013/10/24
===================

- Updated SDK to reflect StackMob REST API changes.
- Updated example to use Sencha Touch 2.3.0.
- [#5] Added support for autoLoading stores.
- [#17] Added in a resetPassword functionality
- [#20] Logout now uses StackMob's API to revoke password.
- [#21] Added a CreateUser functionality.

v0.3.1 - 2012/08/23
===================

- Enhanced usability of the example application.

v0.3 - 2012/06/27
=================

- [#7] Added support for authenticated API calls.
- [#6] Added support for authenticating with StackMob.
- [#8] Added support for manual AJAX requests.

v0.2 - 2012/06/21
=================

- [#1] Added support for paging.
- [#2] Added support for sorting.

v0.1 - 2012/06/19
=============

- Initial version.  Everything has changed!

src/util/CryptoLoader.js
/**
 * Copyright 2012-2013 Palomino Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * A singleton class for loading Jeff Mott's Crypto-JS.  The contents of the load function
 * are copyright Jeff Mott, subject to the license mentioned below.
 *
 * @author Tyler Wolf
 */
Ext.define("Ux.palominolabs.stackmob.util.CryptoLoader", {
    statics: {
        /**
         * Loads the Crypto-JS library
         */
        load: function() {
            /*
             * Crypto-JS v2.5.3
             * http://code.google.com/p/crypto-js/
             * (c) 2009-2012 by Jeff Mott. All rights reserved.
             * http://code.google.com/p/crypto-js/wiki/License
             */
            (typeof Crypto=="undefined"||!Crypto.util)&&function(){var e=window.Crypto={},k=e.util={rotl:function(b,c){return b<<c|b>>>32-c},rotr:function(b,c){return b<<32-c|b>>>c},endian:function(b){if(b.constructor==Number)return k.rotl(b,8)&16711935|k.rotl(b,24)&4278255360;for(var c=0;c<b.length;c++)b[c]=k.endian(b[c]);return b},randomBytes:function(b){for(var c=[];b>0;b--)c.push(Math.floor(Math.random()*256));return c},bytesToWords:function(b){for(var c=[],a=0,i=0;a<b.length;a++,i+=8)c[i>>>5]|=(b[a]&255)<<
                24-i%32;return c},wordsToBytes:function(b){for(var c=[],a=0;a<b.length*32;a+=8)c.push(b[a>>>5]>>>24-a%32&255);return c},bytesToHex:function(b){for(var c=[],a=0;a<b.length;a++)c.push((b[a]>>>4).toString(16)),c.push((b[a]&15).toString(16));return c.join("")},hexToBytes:function(b){for(var c=[],a=0;a<b.length;a+=2)c.push(parseInt(b.substr(a,2),16));return c},bytesToBase64:function(b){if(typeof btoa=="function")return btoa(d.bytesToString(b));for(var c=[],a=0;a<b.length;a+=3)for(var i=b[a]<<16|b[a+1]<<
                8|b[a+2],l=0;l<4;l++)a*8+l*6<=b.length*8?c.push("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".charAt(i>>>6*(3-l)&63)):c.push("=");return c.join("")},base64ToBytes:function(b){if(typeof atob=="function")return d.stringToBytes(atob(b));for(var b=b.replace(/[^A-Z0-9+\/]/ig,""),c=[],a=0,i=0;a<b.length;i=++a%4)i!=0&&c.push(("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".indexOf(b.charAt(a-1))&Math.pow(2,-2*i+8)-1)<<i*2|"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".indexOf(b.charAt(a))>>>
                6-i*2);return c}},e=e.charenc={};e.UTF8={stringToBytes:function(b){return d.stringToBytes(unescape(encodeURIComponent(b)))},bytesToString:function(b){return decodeURIComponent(escape(d.bytesToString(b)))}};var d=e.Binary={stringToBytes:function(b){for(var c=[],a=0;a<b.length;a++)c.push(b.charCodeAt(a)&255);return c},bytesToString:function(b){for(var c=[],a=0;a<b.length;a++)c.push(String.fromCharCode(b[a]));return c.join("")}}}();
            (function(){var e=Crypto,k=e.util,d=e.charenc,b=d.UTF8,c=d.Binary,a=e.SHA1=function(b,l){var f=k.wordsToBytes(a._sha1(b));return l&&l.asBytes?f:l&&l.asString?c.bytesToString(f):k.bytesToHex(f)};a._sha1=function(a){a.constructor==String&&(a=b.stringToBytes(a));var c=k.bytesToWords(a),f=a.length*8,a=[],e=1732584193,g=-271733879,d=-1732584194,j=271733878,m=-1009589776;c[f>>5]|=128<<24-f%32;c[(f+64>>>9<<4)+15]=f;for(f=0;f<c.length;f+=16){for(var p=e,q=g,r=d,s=j,t=m,h=0;h<80;h++){if(h<16)a[h]=c[f+h];else{var n=
                a[h-3]^a[h-8]^a[h-14]^a[h-16];a[h]=n<<1|n>>>31}n=(e<<5|e>>>27)+m+(a[h]>>>0)+(h<20?(g&d|~g&j)+1518500249:h<40?(g^d^j)+1859775393:h<60?(g&d|g&j|d&j)-1894007588:(g^d^j)-899497514);m=j;j=d;d=g<<30|g>>>2;g=e;e=n}e+=p;g+=q;d+=r;j+=s;m+=t}return[e,g,d,j,m]};a._blocksize=16;a._digestsize=20})();
            (function(){var e=Crypto,k=e.util,d=e.charenc,b=d.UTF8,c=d.Binary;e.HMAC=function(a,e,d,f){e.constructor==String&&(e=b.stringToBytes(e));d.constructor==String&&(d=b.stringToBytes(d));d.length>a._blocksize*4&&(d=a(d,{asBytes:!0}));for(var o=d.slice(0),d=d.slice(0),g=0;g<a._blocksize*4;g++)o[g]^=92,d[g]^=54;a=a(o.concat(a(d.concat(e),{asBytes:!0})),{asBytes:!0});return f&&f.asBytes?a:f&&f.asString?c.bytesToString(a):k.bytesToHex(a)}})();

        }
    }
});